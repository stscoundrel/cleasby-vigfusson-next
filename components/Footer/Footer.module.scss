.section {
  padding: 16px 0;
  margin-top: 40px;
  color: white;
  background-color: $color-blue;

  a {
    color: white;
  }
}

.navs {
  display: flex;
  flex-wrap: wrap;
}

.nav {
  flex-basis: 100%;
  margin-bottom: 30px;
}

.navColumns {
  display: flex;
  flex-wrap: wrap;
}

.navColumnItem {
  margin-right: 15px;
  margin-bottom: 15px;
}

.navTitle {
  margin-bottom: 16px;
}

.copyright {
  margin: 30px 0 20px;
  display: block;
  text-align: center;

  a {
    margin-left: 5px;
  }
}

@media only screen and (min-width: 768px) {
  .nav {
    flex-basis: 50%;
  }
}

@media only screen and (min-width: 992px) {
  .nav {
    flex-basis: 33.3%;
  }

  .navs {
    margin-top: 45px;
  }
}