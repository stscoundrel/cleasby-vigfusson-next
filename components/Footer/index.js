import ExternalLink from 'components/ExternalLink'
import LetterLink from 'components/LetterLink'

export default function Footer({ letters }) {
  return (
    <footer>
      <hr />

      <section>
        <h2>About</h2>
        <p>Based on Cleasby & Vigfusson Dictionary Old Norse dictionary.</p>
        <p><em>Icelandic-English</em> dictionary was started by Richard Cleasby and
        finished by Gudbrand Vigfusson.</p>
        <p>It was published in 1874,
        which leads to there being many public domain versions of the book available.</p>
      </section>

      <nav>
        <p>Dictionary project</p>
        <ul>
          <li>
            <ExternalLink
              title="Source code"
              href="https://github.com/stscoundrel/cleasby-vigfusson-next"
            />
          </li>
          <li>
            <ExternalLink
              title="Data source"
              href="https://github.com/stscoundrel/cleasby-vigfusson-dictionary"
            />
          </li>
        </ul>
      </nav>

      <nav>
        <p>Quick links</p>
        <ul>
          {letters.map((entry) => (
            <li key={entry.slug}>
              <LetterLink letter={entry} />
            </li>
          ))}
        </ul>
      </nav>

      <small>Copyright Â© 2020 Sampo Silvennoinen / StScoundrel</small>
    </footer>
  )
}
