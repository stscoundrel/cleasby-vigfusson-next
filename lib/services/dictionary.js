const { getDictionary } = require('cleasby-vigfusson-dictionary');
const slugify = require('slugify');

const addSlugs = (words) => {
  const existingSlugs = {};

  const formattedWords = words.map((word) => {
    let slug = slugify(word.word).toLowerCase();

    if (existingSlugs[slug]) {
      // Double slug, make unique.
      existingSlugs[slug] += 1;
      slug = `${slug}-${existingSlugs[slug]}`;
    } else {
      existingSlugs[slug] = 1;
    }

    return {
      ...word,
      slug,
    };
  });

  return formattedWords;
};

export const getAllWords = () => {
  const words = getDictionary();

  /**
   * Add URL safe slugs.
   */
  const formattedWords = addSlugs(words);

  return formattedWords;
};

export default {
  getAllWords,
};
