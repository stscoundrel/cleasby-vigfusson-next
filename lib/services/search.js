export const searchDictionary = (search, dictionary) => {
  const results = dictionary.filter((entry) => {
    let matchesSearch = false

    if (entry.word.includes(search)) {
      matchesSearch = true
    }

    if (entry.slug.includes(search)) {
      matchesSearch = true
    }

    entry.definitions.forEach((definition) => {
      if (definition.includes(search)) {
        matchesSearch = true
      }
    })

    return matchesSearch
  })

  return results
}

export default {
  searchDictionary,
}
