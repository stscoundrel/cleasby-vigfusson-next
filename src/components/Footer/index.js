import ExternalLink from 'components/ExternalLink'
import LetterLink from 'components/LetterLink'
import ContentArea from 'components/ContentArea'
import styles from './Footer.module.scss'

export default function Footer({ letters }) {
  return (
    <footer className={styles.section}>
      <div className="container">

        <ContentArea>
          <h2>About</h2>
          <p>Based on Cleasby & Vigfusson Dictionary Old Norse dictionary.</p>
          <p><em>Icelandic-English</em> dictionary was started by Richard Cleasby and
          finished by Gudbrand Vigfusson.</p>
          <p>It was published in 1874,
          which leads to there being many public domain versions of the book available.</p>
        </ContentArea>

        <ContentArea>
          <h3>Old Norse language</h3>
          <p>Old Norse was a North Germanic language that was spoken by inhabitants of
          Scandinavia and their overseas settlements from about the 7th to the 15th centuries.</p>

          <p>Also known as &quot;the viking language&quot;,
          &quot;Old Nordic&quot;, or
          &quot;Old Scandinavian&quot;</p>
        </ContentArea>

        <div className={styles.navs}>
          <nav className={styles.nav}>
            <h4 className={styles.navTitle}>Dictionary project</h4>
            <ul>
              <li>
                <ExternalLink
                  title="Source code"
                  href="https://github.com/stscoundrel/cleasby-vigfusson-next"
                />
              </li>
              <li>
                <ExternalLink
                  title="Data source"
                  href="https://github.com/stscoundrel/cleasby-vigfusson-dictionary"
                />
              </li>
              <li>
                <ExternalLink
                  title="Abbreviations"
                  href="https://github.com/stscoundrel/cleasby-vigfusson-abbreviations"
                />
              </li>
            </ul>
          </nav>

          <nav className={styles.nav}>
            <h4 className={styles.navTitle}>Related packages</h4>
            <ul>
              <li>
                <ExternalLink
                  title="Younger Futhark Runes"
                  href="https://github.com/stscoundrel/younger-futhark"
                />
              </li>
              <li>
                <ExternalLink
                  title="Old Norse Alphabet"
                  href="https://github.com/stscoundrel/old-norse-alphabet"
                />
              </li>
              <li>
                <ExternalLink
                  title="Old Norse Alphabet Sort"
                  href="https://github.com/stscoundrel/old-norse-alphabet-sort"
                />
              </li>
              <li>
                <ExternalLink
                  title="Gangleri"
                  href="https://github.com/stscoundrel/gangleri"
                />
              </li>
              <li>
                <ExternalLink
                  title="Teljari"
                  href="https://github.com/stscoundrel/teljari"
                />
              </li>
              <li>
                <ExternalLink
                  title="Markari"
                  href="https://github.com/stscoundrel/markari"
                />
              </li>
              <li>
                <ExternalLink
                  title="Spyrjari"
                  href="https://github.com/stscoundrel/spyrjari"
                />
              </li>

            </ul>
          </nav>

          <nav className={styles.nav}>
            <h4 className={styles.navTitle}>Quick links</h4>
            <ul className={styles.navColumns}>
              {letters.map((entry) => (
                <li className={styles.navColumnItem} key={entry.slug}>
                  <LetterLink letter={entry} />
                </li>
              ))}
            </ul>
          </nav>
        </div>
        <small className={styles.copyright}>Copyright Â© 2020
          <ExternalLink
            title="Sampo Silvennoinen / StScoundrel"
            href="https://github.com/stscoundrel"
          />
        </small>
      </div>
    </footer>
  )
}
