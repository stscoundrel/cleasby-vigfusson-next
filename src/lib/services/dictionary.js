const { getDictionary } = require('cleasby-vigfusson-dictionary')
const { VALID_AS_FIRST } = require('old-norse-alphabet')
const { slugifyWord, slugifyLetter } = require('../utils/slugs')

let cachedDictionary = null

const addSlugs = (words) => {
  const existingSlugs = {}

  const formattedWords = words.map((word) => {
    let slug = slugifyWord(word.word).toLowerCase()

    if (slug.charAt(0) === '-') {
      slug = slug.substring(1)
    }

    if (existingSlugs[slug]) {
      // Double slug, make unique.
      existingSlugs[slug] += 1
      slug = `${slug}-${existingSlugs[slug]}`
    } else {
      existingSlugs[slug] = 1
    }

    return {
      ...word,
      slug,
    }
  })

  return formattedWords
}

export const getAllWords = () => {
  if (cachedDictionary) return cachedDictionary

  const words = getDictionary()

  /**
   * Add URL safe slugs.
   */
  const formattedWords = addSlugs(words)

  cachedDictionary = formattedWords

  return formattedWords
}

export const getByLetter = (letter) => {
  const words = getAllWords()
  const byLetter = words.filter((entry) => (
    entry.word.charAt(0).toLowerCase() === letter.toLowerCase()))

  return byLetter
}

export const getWord = (word) => getAllWords().filter((entry) => entry.slug === word)[0]

export const getAlphabet = () => {
  const letters = [...VALID_AS_FIRST.filter((letter) => letter !== 'ǫ' && letter !== 'ø'), 'ö']

  const formattedLetters = letters.map((letter) => ({
    letter,
    slug: slugifyLetter(letter),
  }))

  return formattedLetters
}

export default {
  getAllWords,
  getByLetter,
  getWord,
  getAlphabet,
}
