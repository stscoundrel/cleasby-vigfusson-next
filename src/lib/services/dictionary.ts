import { getDictionary, DictionaryEntry as RawDictionaryEntry } from 'cleasby-vigfusson-dictionary'
import { VALID_AS_FIRST } from 'old-norse-alphabet'
import { oldNorseSort } from 'old-norse-alphabet-sort'
import { slugifyWord, slugifyLetter } from '../utils/slugs'

export interface DictionaryEntry extends RawDictionaryEntry {
  slug: string,
}

export interface DictionaryEntryDTO {
  word: string,
  slug: string
}

export interface AlphabetLetter {
  letter: string,
  slug: string
}

let cachedDictionary: DictionaryEntry[] | null = null

const addSlugs = (words: RawDictionaryEntry[]): DictionaryEntry[] => {
  const existingSlugs = {}

  const formattedWords = words.map((word) => {
    let slug = slugifyWord(word.word).toLowerCase()

    if (slug.charAt(0) === '-') {
      slug = slug.substring(1)
    }

    if (existingSlugs[slug]) {
      // Double slug, make unique.
      existingSlugs[slug] += 1
      slug = `${slug}-${existingSlugs[slug]}`
    } else {
      existingSlugs[slug] = 1
    }

    return {
      ...word,
      slug,
    }
  })

  return formattedWords
}

export const getAllWords = (): DictionaryEntry[] => {
  if (cachedDictionary) return cachedDictionary

  const words = getDictionary()

  /**
   * Add URL safe slugs.
   */
  const formattedWords = addSlugs(words)

  /**
   * Ensure headwords are lowercased.
   * Original source has some in all caps.
   * That is true to the layout of the book,
   * but makes little sense in website.
   */
  const dictionary = formattedWords.map((entry) => ({ ...entry, word: entry.word.toLowerCase() }))

  cachedDictionary = dictionary

  return dictionary
}

export const getByLetter = (letter: string): DictionaryEntryDTO[] => {
  const words = getAllWords()
  const byLetter = words
    .filter((entry) => (
      entry.word.charAt(0).toLowerCase() === letter.toLowerCase()))
    .map((entry) => {
      // Simpler DTO dictionary entry.
      const { word, slug } = entry
      return { word, slug }
    })
    .sort((a, b) => oldNorseSort(a.word, b.word))

  return byLetter
}

export const getWord = (slug: string): DictionaryEntry => (
  getAllWords().filter((entry) => entry.slug === slug)[0]
)

export const getRandomEntries = (): DictionaryEntry[] => (
  // Return entries fit to be randomized "teasers"
  // Therefore, content should be short, but not too short.
  getAllWords()
    .sort(() => Math.random() - 0.5)
    .filter((entry) => entry.definitions[0].length < 50 && entry.definitions[0].length > 15)
    .slice(0, 36)
    .sort((a, b) => oldNorseSort(a.word, b.word))
)

export const getSimilarWords = (entry: DictionaryEntry) : DictionaryEntry[] => getAllWords()
  .filter((dEntry) => dEntry.word.toLocaleLowerCase() === entry.word.toLocaleLowerCase()
  && dEntry.slug !== entry.slug)

export const getAlphabet = (): AlphabetLetter[] => {
  const letters = [...VALID_AS_FIRST.filter((letter) => letter !== 'ǫ' && letter !== 'ø'), 'ö']

  const formattedLetters = letters.map((letter) => ({
    letter,
    slug: slugifyLetter(letter),
  }))

  return formattedLetters
}

export default {
  getAllWords,
  getByLetter,
  getWord,
  getRandomEntries,
  getAlphabet,
}
