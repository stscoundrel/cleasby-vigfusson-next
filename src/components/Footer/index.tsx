import ExternalLink from 'components/ExternalLink'
import LetterLink from 'components/LetterLink'
import ContentArea from 'components/ContentArea'
import { AlphabetLetter } from 'lib/services/dictionary'
import styles from './Footer.module.scss'

interface FooterProps{
  letters: AlphabetLetter[]
}

export default function Footer({ letters }: FooterProps) {
  return (
    <footer className={styles.section}>
      <div className="container">

        <ContentArea>
          <h2>About</h2>
          <p>Based on Cleasby & Vigfusson Dictionary Old Norse dictionary.</p>
          <p><em>Icelandic-English</em> dictionary was started by Richard Cleasby and
          finished by Gudbrand Vigfusson.</p>
          <p>It was published in 1874,
          which leads to there being many public domain versions of the book available.</p>
        </ContentArea>

        <ContentArea>
          <h3>Old Norse language</h3>
          <p>Old Norse was a North Germanic language that was spoken by inhabitants of
          Scandinavia and their overseas settlements from about the 7th to the 15th centuries.</p>

          <p>Also known as &quot;the viking language&quot;,
          &quot;Old Nordic&quot;, or
          &quot;Old Scandinavian&quot;</p>
        </ContentArea>

        <div className={styles.navs}>
          <nav className={styles.nav}>
            <h4 className={styles.navTitle}>Dictionary project</h4>
            <ul>
              <li>
                <ExternalLink
                  title="Source code"
                  href="https://github.com/stscoundrel/cleasby-vigfusson-next"
                />
              </li>
              <li>
                <ExternalLink
                  title="Data source"
                  href="https://github.com/stscoundrel/cleasby-vigfusson-dictionary"
                />
              </li>
              <li>
                <ExternalLink
                  title="Abbreviations"
                  href="https://github.com/stscoundrel/cleasby-vigfusson-abbreviations"
                />
              </li>
              <li>
                <ExternalLink
                  title="Younger Futhark Runes"
                  href="https://github.com/stscoundrel/younger-futhark"
                />
              </li>
            </ul>
          </nav>

          <nav className={styles.nav}>
            <h4 className={styles.navTitle}>Related dictionary projects</h4>
            <ul>
              <li>
                <ExternalLink
                  title="A Concise Dictionary of Old Icelandic"
                  href="https://old-icelandic.vercel.app/"
                />
              </li>
              <li>
                <ExternalLink
                  title="K.F Söderwall's Old Swedish Dictionary"
                  href="https://old-swedish-dictionary.vercel.app/"
                />
              </li>
              <li>
                <ExternalLink
                  title="Dictionary of the Old Norwegian Language"
                  href="https://old-norwegian-dictionary.vercel.app/"
                />
              </li>
              <li>
                <ExternalLink
                  title="Otto Kalkar's Old Danish Dictionary"
                  href="https://old-danish-dictionary.vercel.app/"
                />
              </li>
            </ul>
          </nav>

          <nav className={styles.nav}>
            <h4 className={styles.navTitle}>Quick links</h4>
            <ul className={styles.navColumns}>
              {letters.map((entry) => (
                <li className={styles.navColumnItem} key={entry.slug}>
                  <LetterLink letter={entry} />
                </li>
              ))}
            </ul>
          </nav>
        </div>
        <small className={styles.copyright}>{`Copyright © 2020 - ${new Date().getFullYear()}`}
          <br />
          <ExternalLink
            title="Sampo Silvennoinen / StScoundrel"
            href="https://github.com/stscoundrel"
          />
        </small>
      </div>
    </footer>
  )
}
