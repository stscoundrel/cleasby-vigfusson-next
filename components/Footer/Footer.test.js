import ReactDOM from 'react-dom'
import Footer from 'components/Footer'
import renderer from 'react-test-renderer'
import { getAlphabet } from 'lib/services/dictionary'

const letters = getAlphabet()

describe('Footer component', () => {
  test('Does not crash', () => {
    const div = document.createElement('div')
    ReactDOM.render(<Footer letters={letters}/>, div)
    ReactDOM.unmountComponentAtNode(div)
  })

  test('Matches snapshot', () => {
    const tree = renderer.create(<Footer letters={letters} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
