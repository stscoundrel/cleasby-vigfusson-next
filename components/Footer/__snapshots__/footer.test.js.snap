// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer component Matches snapshot 1`] = `
<footer>
  <hr />
  <section>
    <h2>
      About
    </h2>
    <p>
      Based on Cleasby & Vigfusson Dictionary Old Norse dictionary.
    </p>
    <p>
      <em>
        Icelandic-English
      </em>
       dictionary was started by Richard Cleasby and finished by Gudbrand Vigfusson.
    </p>
    <p>
      It was published in 1874, which leads to there being many public domain versions of the book available.
    </p>
  </section>
  <nav>
    <p>
      Dictionary project
    </p>
    <ul>
      <li>
        <a
          href="https://github.com/stscoundrel/cleasby-vigfusson-next"
          rel="noopener noreferrer"
          target="_blank"
        >
          Source code
        </a>
      </li>
      <li>
        <a
          href="https://github.com/stscoundrel/cleasby-vigfusson-dictionary"
          rel="noopener noreferrer"
          target="_blank"
        >
          Data source
        </a>
      </li>
    </ul>
  </nav>
  <nav>
    <p>
      Quick links
    </p>
    <ul>
      <li>
        <a
          href="/letter/a"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          a
        </a>
      </li>
      <li>
        <a
          href="/letter/a2"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          á
        </a>
      </li>
      <li>
        <a
          href="/letter/b"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          b
        </a>
      </li>
      <li>
        <a
          href="/letter/d"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          d
        </a>
      </li>
      <li>
        <a
          href="/letter/e"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          e
        </a>
      </li>
      <li>
        <a
          href="/letter/e2"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          é
        </a>
      </li>
      <li>
        <a
          href="/letter/f"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          f
        </a>
      </li>
      <li>
        <a
          href="/letter/g"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          g
        </a>
      </li>
      <li>
        <a
          href="/letter/h"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          h
        </a>
      </li>
      <li>
        <a
          href="/letter/i"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          i
        </a>
      </li>
      <li>
        <a
          href="/letter/i2"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          í
        </a>
      </li>
      <li>
        <a
          href="/letter/j"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          j
        </a>
      </li>
      <li>
        <a
          href="/letter/k"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          k
        </a>
      </li>
      <li>
        <a
          href="/letter/l"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          l
        </a>
      </li>
      <li>
        <a
          href="/letter/m"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          m
        </a>
      </li>
      <li>
        <a
          href="/letter/n"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          n
        </a>
      </li>
      <li>
        <a
          href="/letter/o"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          o
        </a>
      </li>
      <li>
        <a
          href="/letter/o2"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          ó
        </a>
      </li>
      <li>
        <a
          href="/letter/p"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          p
        </a>
      </li>
      <li>
        <a
          href="/letter/r"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          r
        </a>
      </li>
      <li>
        <a
          href="/letter/s"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          s
        </a>
      </li>
      <li>
        <a
          href="/letter/t"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          t
        </a>
      </li>
      <li>
        <a
          href="/letter/u"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          u
        </a>
      </li>
      <li>
        <a
          href="/letter/u2"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          ú
        </a>
      </li>
      <li>
        <a
          href="/letter/v"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          v
        </a>
      </li>
      <li>
        <a
          href="/letter/y"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          y
        </a>
      </li>
      <li>
        <a
          href="/letter/y2"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          ý
        </a>
      </li>
      <li>
        <a
          href="/letter/th"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          þ
        </a>
      </li>
      <li>
        <a
          href="/letter/ae"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          æ
        </a>
      </li>
      <li>
        <a
          href="/letter/oe3"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          œ
        </a>
      </li>
      <li>
        <a
          href="/letter/oe"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          ö
        </a>
      </li>
    </ul>
  </nav>
  <small>
    Copyright © 2020 Sampo Silvennoinen / StScoundrel
  </small>
</footer>
`;
